### ISSUE TITLE
bcrypt.compare() always return false even when it´s supposed to be true

### ISSUE THREAD

**Issue by s-pl** [kelektiv/node.bcrypt.js#1037]
Hello, I'm working on an authentication system in MongoDB. When storing passwords, everything seems fine:
{"_id":{"$oid":"66408f417315a786f0d1d279"},"username":"test","email":"test@mail.com","password":"$2b$10$e2KuxFiAq4oVl7DaL80TX.9udp65K9uoiVOSfXZNmzHP8rVqIR5bG","role":"user","createdAt":{"$date":{"$numberLong":"1715507009424"}},"updatedAt":{"$date":{"$numberLong":"1715507009424"}},"__v":{"$numberInt":"0"}}
But then, when comparing the hash with the password, it always returns incorrect (even when it's correct). This is the method I'm using:

```
userSchema.pre('save', async function (next) {
  const user = this;
  if (!user.isModified('password')) return next();

  try {
    const salt = await bcrypt.genSalt();
    user.password = await bcrypt.hash(user.password, salt);
    next();
  } catch (error) {
    return next(error);
  }
});

userSchema.methods.comparePassword = async function (password) {
    
    console.log(password,this.password)
    return bcrypt.compare(password, this.password);
  
};
```

And this is the comparison log:
**test** **$2b$10$e2KuxFiAq4oVl7DaL80TX.9udp65K9uoiVOSfXZNmzHP8rVqIR5bG**

I think there's an error in the `bcrypt.compare` function, but I'm very lost

**Comment by parth-develops** [comment]
@s-pl yes i have been facing this issue as well with both bcryptjs and bcrypt

**Comment by wiryonolau** [comment]
any update, I got similar problem. 
Check using php password_verify return true, but in js always false

**Comment by Talhaayubkhan** [comment]
similar problem, anyone who solve this???
the password even correct but in login condiotion it give an error that password is invalid???

**Comment by Talhaayubkhan** [comment]
i have face the same issue?? anyone who solve it??

**Comment by Talhaayubkhan** [comment]
> Hello, I'm working on an authentication system in MongoDB. When storing passwords, everything seems fine: `{"_id":{"$oid":"66408f417315a786f0d1d279"},"username":"test","email":"test@mail.com","password":"$2b$10$e2KuxFiAq4oVl7DaL80TX.9udp65K9uoiVOSfXZNmzHP8rVqIR5bG","role":"user","createdAt":{"$date":{"$numberLong":"1715507009424"}},"updatedAt":{"$date":{"$numberLong":"1715507009424"}},"__v":{"$numberInt":"0"}}` But then, when comparing the hash with the password, it always returns incorrect (even when it's correct). This is the method I'm using:
> 
> ```
> userSchema.pre('save', async function (next) {
>   const user = this;
>   if (!user.isModified('password')) return next();
> 
>   try {
>     const salt = await bcrypt.genSalt();
>     user.password = await bcrypt.hash(user.password, salt);
>     next();
>   } catch (error) {
>     return next(error);
>   }
> });
> 
> userSchema.methods.comparePassword = async function (password) {
>     
>     console.log(password,this.password)
>     return bcrypt.compare(password, this.password);
>   
> };
> ```
> 
> And this is the comparison log: **test** **$2b$10$e2KuxFiAq4oVl7DaL80TX.9udp65K9uoiVOSfXZNmzHP8rVqIR5bG**
> 
> I think there's an error in the `bcrypt.compare` function, but I'm very lost

**Comment by Talhaayubkhan** [comment]
> Hello, I'm working on an authentication system in MongoDB. When storing passwords, everything seems fine: `{"_id":{"$oid":"66408f417315a786f0d1d279"},"username":"test","email":"test@mail.com","password":"$2b$10$e2KuxFiAq4oVl7DaL80TX.9udp65K9uoiVOSfXZNmzHP8rVqIR5bG","role":"user","createdAt":{"$date":{"$numberLong":"1715507009424"}},"updatedAt":{"$date":{"$numberLong":"1715507009424"}},"__v":{"$numberInt":"0"}}` But then, when comparing the hash with the password, it always returns incorrect (even when it's correct). This is the method I'm using:
> 
> ```
> userSchema.pre('save', async function (next) {
>   const user = this;
>   if (!user.isModified('password')) return next();
> 
>   try {
>     const salt = await bcrypt.genSalt();
>     user.password = await bcrypt.hash(user.password, salt);
>     next();
>   } catch (error) {
>     return next(error);
>   }
> });
> 
> userSchema.methods.comparePassword = async function (password) {
>     
>     console.log(password,this.password)
>     return bcrypt.compare(password, this.password);
>   
> };
> ```
> 
> And this is the comparison log: **test** **$2b$10$e2KuxFiAq4oVl7DaL80TX.9udp65K9uoiVOSfXZNmzHP8rVqIR5bG**
> 
> I think there's an error in the `bcrypt.compare` function, but I'm very lost

//my solution!!
ok guys! i have find the bug!!
if you set select:false in password (user model),
try to remove select("-password") or ("+password") in user auth model (loginUser)

**Comment by Talhaayubkhan** [comment]
> @s-pl yes i have been facing this issue as well with both bcryptjs and bcrypt

// my solution
can u try this, if you set select:false in password (user model),
try to remove select("-password") or ("+password") in user auth model (loginUser)

**Comment by M00N15** [comment]
I am using postgresSQL. But its not a concern here i guess. facing the same issue as above. will it help if i am using a saltRound of lower value? it obviously means reducing security. but i think the length of the hashed password generated is the problem here. what do you guys think?

**Comment by s-pl** [comment]
the common solution is to extend the maximum lenght of characters that your
db can handle. But if that doesn’t work try to install the previous version
of Bcrypt.

**Comment by WXZhang365** [comment]
> I am using postgresSQL. But its not a concern here i guess. facing the same issue as above. will it help if i am using a saltRound of lower value? it obviously means reducing security. but i think the length of the hashed password generated is the problem here. what do you guys think?

@M00N15 
I try to use sequelize
```
@Column(DataType.CHAR(255,true)) 
```
binary true option.
when i use this option , the compare result is true.
I think mb because sequelize and postgresql have some bug.

**Comment by WXZhang365** [comment]
> > I am using postgresSQL. But its not a concern here i guess. facing the same issue as above. will it help if i am using a saltRound of lower value? it obviously means reducing security. but i think the length of the hashed password generated is the problem here. what do you guys think?
> 
> @M00N15 I try to use sequelize
> 
> ```
> @Column(DataType.CHAR(255,true)) 
> ```
> 
> binary true option. when i use this option , the compare result is true. I think mb because sequelize and postgresql have some bug.

@M00N15 

I think I know why.
```
 const p = bcrypt.hashSync(loginFrom.password, 10);
    const result = await UserModel.create({
      username: loginFrom.username,
      password: p,
    });
    await result.reload({
      attributes: ['password'],
    });
    console.log(p == result.password);
    console.log(Buffer.from(p), Buffer.from(p).length);
    console.log(
      Buffer.from(result.password),
      Buffer.from(result.password).length,
    );
```
result is :
```
<Buffer 24 32 62 24 31 30 24 37 54 43 4b 32 4f 71 71 4d 55 50 48 59 62 42 49 47 59 30 53 63 75 30 73 66 75 35 6e 4a 33 32 66 5a 62 43 78 78 47 75 46 61 4b 58 ... 10 more bytes> 60
<Buffer 24 32 62 24 31 30 24 37 54 43 4b 32 4f 71 71 4d 55 50 48 59 62 42 49 47 59 30 53 63 75 30 73 66 75 35 6e 4a 33 32 66 5a 62 43 78 78 47 75 46 61 4b 58 ... 205 more bytes> 255
```

my colume is 
```
  @Column(DataType.CHAR(255, false))
  password: string;
```

**Comment by Nana-Dilan-kenye** [comment]
Use this   await user.updateOne({ $set: { password: hashedPassword } });

**Comment by parth-develops** [comment]
> > @s-pl yes i have been facing this issue as well with both bcryptjs and bcrypt
> 
> // my solution can u try this, if you set select:false in password (user model), try to remove select("-password") or ("+password") in user auth model (loginUser)

For me it was a logical error which is fixed now, bcrypt is working fine.

**Comment by ArcherFlesk** [comment]
Yeah, dealing with the same issue as well. Tried regressing the package version to 5.1.0, adjusting variable data types and lengths with no luck. Checked my logs whenever creating a fake user account ensuring the hash was the same both putting and pulling out from the database.

**Comment by WXZhang365** [comment]
> Yeah, dealing with the same issue as well. Tried regressing the package version to 5.1.0, adjusting variable data types and lengths with no luck. Checked my logs whenever creating a fake user account ensuring the hash was the same both putting and pulling out from the database.

I don't know why, but sometimes is normal in other projects.

**Comment by ArcherFlesk** [comment]
> > Yeah, dealing with the same issue as well. Tried regressing the package version to 5.1.0, adjusting variable data types and lengths with no luck. Checked my logs whenever creating a fake user account ensuring the hash was the same both putting and pulling out from the database.
> 
> I don't know why, but sometimes is normal in other projects.

Any correlations or ideas that allow it to work for some projects, but not others?

**Comment by greenlife-developer** [comment]
In my case, the code works, but it fails anytime and I have to reset the password. So it works again, then after sometimes it fails again. What solution is there to this, is there something specific we need to do?

**Comment by D50000** [comment]
Same issue here, it always return `true`
`const isPasswordValid = bcrypt.compareSync(
          password as string,
          encryptPWD,
        );`
